File: stations/_archive/hook-observatory-py/handlers/issue_sync.py
"""
PostToolUse/Bash: auto-detect issue refs in git commits and sync to GitHub.

When a Bash command contains 'git commit', checks the latest commit for
issue references (#N, Fixes #N, Closes #N, Part of #N) and posts a
progress comment on the referenced GitHub Issue.

Also checks the current branch name for issue references (feature/foo-#42).
"""

from __future__ import annotations

import re

from .base import ALLOW, HookResult, message, run_cmd

REPO = "JonesHong/workshop"


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """Post commit info to referenced GitHub Issues."""
    command = tool_input.get("command", "")

    # Only trigger on git commit commands
    if "git commit" not in command:
        return ALLOW

    # Get the latest commit message + hash
    log_result = run_cmd(["git", "log", "-1", "--format=%h%n%s"], timeout=5)
    if log_result is None or log_result.returncode != 0:
        return ALLOW

    lines = log_result.stdout.strip().split("\n", 1)
    if len(lines) < 2:
        return ALLOW

    commit_hash = lines[0]
    commit_subject = lines[1]

    # Extract issue references from commit message
    full_msg_result = run_cmd(["git", "log", "-1", "--format=%s%n%b"], timeout=5)
    full_msg = full_msg_result.stdout.strip() if full_msg_result else commit_subject

    issue_numbers: set[str] = set()

    # Match: #N, Closes #N, Fixes #N, Refs #N, Part of #N
    for m in re.finditer(r"(?:Closes|Fixes|Refs|Part of)\s+#(\d+)", full_msg):
        issue_numbers.add(m.group(1))
    # Match standalone #N (not preceded by color hex like #FFFFFF)
    for m in re.finditer(r"(?<!\w)#(\d+)\b", full_msg):
        issue_numbers.add(m.group(1))

    # Also check branch name for issue reference (feature/slug-#42)
    branch_result = run_cmd(["git", "branch", "--show-current"], timeout=5)
    if branch_result and branch_result.returncode == 0:
        branch = branch_result.stdout.strip()
        branch_match = re.search(r"#(\d+)", branch)
        if branch_match:
            issue_numbers.add(branch_match.group(1))

    if not issue_numbers:
        return ALLOW

    # Post comment on each referenced issue (fire-and-forget, don't block)
    synced = []
    for num in issue_numbers:
        comment = f"Commit `{commit_hash}`: {commit_subject}"
        result = run_cmd(
            ["gh", "issue", "comment", num, "--repo", REPO, "--body", comment],
            timeout=10,
        )
        if result and result.returncode == 0:
            synced.append(f"#{num}")

    if synced:
        return message(f"Synced commit {commit_hash} to {', '.join(synced)}")

    return ALLOW

File: internal/handlers/issue_sync.go
package handlers

// issue_sync.go — Go port of handlers/issue_sync.py
//
// PostToolUse/Bash: auto-detect issue refs in git commits and sync to GitHub.
//
// When a Bash command contains 'git commit', checks the latest commit for
// issue references (#N, Fixes #N, Closes #N, Part of #N) and posts a
// progress comment on the referenced GitHub Issue.
//
// Matcher: "Bash", deferrable.

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/joneshong/hook-observatory/internal/clients"
	"github.com/joneshong/hook-observatory/internal/core"
)

const issueSyncRepo = "JonesHong/workshop"

var (
	// Match: Closes/Fixes/Refs/Part of #N
	reIssueKeyword = regexp.MustCompile(`(?i)(?:Closes|Fixes|Refs|Part of)\s+#(\d+)`)
	// Match standalone #N (not preceded by word char — avoids hex colors like #FF0000)
	reIssueStandalone = regexp.MustCompile(`(?:^|[^\w])#(\d+)\b`)
	// Match #N in branch name
	reIssueBranch = regexp.MustCompile(`#(\d+)`)
)

// NOTE: Not registered. Python handlers/__init__.py does not import or register
// issue_sync — it's inactive in production. Keeping the implementation for
// future opt-in, but no core.Register() call to preserve parity.
func init() {}

func issueSyncHandle(_, _ string, toolInput map[string]any, _ string) core.HookResult {
	command, _ := toolInput["command"].(string)
	if !strings.Contains(command, "git commit") {
		return core.Allow()
	}

	// Get latest commit hash + subject
	gitLog := core.RunCmd([]string{"git", "log", "-1", "--format=%h%n%s"}, "", 5*time.Second, "")
	if gitLog == nil || gitLog.ExitCode != 0 {
		return core.Allow()
	}
	parts := strings.SplitN(strings.TrimSpace(gitLog.Stdout), "\n", 2)
	if len(parts) < 2 {
		return core.Allow()
	}
	commitHash := parts[0]
	commitSubject := parts[1]

	// Get full message (subject + body)
	fullMsgOut := core.RunCmd([]string{"git", "log", "-1", "--format=%s%n%b"}, "", 5*time.Second, "")
	fullMsg := commitSubject
	if fullMsgOut != nil && fullMsgOut.ExitCode == 0 {
		fullMsg = strings.TrimSpace(fullMsgOut.Stdout)
	}

	// Collect issue numbers
	issueNumbers := make(map[string]struct{})

	for _, m := range reIssueKeyword.FindAllStringSubmatch(fullMsg, -1) {
		issueNumbers[m[1]] = struct{}{}
	}
	for _, m := range reIssueStandalone.FindAllStringSubmatch(fullMsg, -1) {
		issueNumbers[m[1]] = struct{}{}
	}

	// Check branch name
	branchOut := core.RunCmd([]string{"git", "branch", "--show-current"}, "", 5*time.Second, "")
	if branchOut != nil && branchOut.ExitCode == 0 {
		branch := strings.TrimSpace(branchOut.Stdout)
		if m := reIssueBranch.FindStringSubmatch(branch); m != nil {
			issueNumbers[m[1]] = struct{}{}
		}
	}

	if len(issueNumbers) == 0 {
		return core.Allow()
	}

	// Post comment on each referenced issue (fire-and-forget)
	var synced []string
	comment := fmt.Sprintf("Commit `%s`: %s", commitHash, commitSubject)
	for num := range issueNumbers {
		_, err := clients.RunGH([]string{
			"issue", "comment", num,
			"--repo", issueSyncRepo,
			"--body", comment,
		})
		if err == nil {
			synced = append(synced, "#"+num)
		}
	}

	if len(synced) > 0 {
		return core.Message(fmt.Sprintf("Synced commit %s to %s", commitHash, strings.Join(synced, ", ")))
	}
	return core.Allow()
}
